src/functional/geometry/convex_hull.zig

```diff
@@ -0,0 +1,167 @@
+// Copyright 2024 TerseTS Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Implementation of the Graham's scan which can maintain a convex hull from the paper
+//! "Graham, R.L. An Efficient Algorithm for Determining the Convex Hull of a Finite Planar Set.
+//! Information Processing Letters. 1 (4): 132–133. 1972.
+//! https://doi.org/10.1016/0020-0190(72)90045-2.
+//! Convex Hulls are used by the Slide Filter and Optimal PLA algorithms implemented in
+//! `src/functional/swing_slide_filter.zig` and `src/functional/optimal_pla.zig`.
+
+const std = @import("std");
+const mem = std.mem;
+const ArrayList = std.ArrayList;
+const testing = std.testing;
+
+const tersets = @import("../../tersets.zig");
+const Point = tersets.Point;
+
+/// Enum for the angle's `Turn` of three consecutive points A, B, and C. Essentially, it describes
+/// whether the path from A to B to C makes a `left` turn, a `right` turn, or continues in a
+/// straight line also called collinear.
+const Turn = enum(i8) { right, left, collinear };
+
+/// Convex Hull formed by an upper and lower hull.
+pub const ConvexHull = struct {
+    lower_hull: ArrayList(Point),
+    upper_hull: ArrayList(Point),
+
+    // Initialize the container with a given `allocator`.
+    pub fn init(allocator: mem.Allocator) !ConvexHull {
+        return ConvexHull{
+            .lower_hull = ArrayList(Point).init(allocator),
+            .upper_hull = ArrayList(Point).init(allocator),
+        };
+    }
+
+    // Deinitialize the container and free the allocated memory.
+    pub fn deinit(self: *ConvexHull) void {
+        self.lower_hull.deinit();
+        self.upper_hull.deinit();
+    }
+
+    /// Add new `point` to the convex hull.
+    pub fn addPoint(self: *ConvexHull, point: Point) !void {
+        try addPointToHull(&self.upper_hull, Turn.right, point);
+        try addPointToHull(&self.lower_hull, Turn.left, point);
+    }
+};
+
+/// Auxiliary function to add a new `point` to a given `hull` of the convex hull. The function uses
+/// the given `turn` to correctly add the new point.
```

Comment:
Maybe add another sentence to the comment making it clear why `addPointToHull()` is not part of the `ConvexHull` despite being a helper functions for methods in that type. Seems it should be possible to implement it as a private method that does not take `self` as the first argument like `ArrayList.init()` as shown in the Zig documentation [here](https://ziglang.org/documentation/master/#struct).

Reply:
Well, I just tried declaring the function inside the structure without `self` as the first argument and everything works well. So I fixed this comment by bringing the function inside and slightly changing the documentation.